File: basic/minimaxtree/MinimaxTree.js
const input = require('fs').readFileSync('/dev/stdin').toString().split('\n');
const [N, R] = input[0].split(" ").map(Number);

const nodeValues = Array(N + 1).fill(-1);
const graph = Array(N + 1).fill().map(() => []);

// Build the graph
for (let i = 1; i < N; i++) {
    const [u, v] = input[i].split(" ").map(Number);
    graph[u].push(v);
    graph[v].push(u);
}

// Read leaf values
const L = Number(input[N]);
for (let i = 0; i < L; i++) {
    const [leafId, leafValue] = input[N + 1 + i].split(" ").map(Number);
    nodeValues[leafId] = leafValue;
}

// DFS function
function dfs(currentNode, parentNode, depth) {
    if (nodeValues[currentNode] >= 0) return nodeValues[currentNode];

    if (depth % 2 === 0) {
        let bestValue = 0;
        for (const child of graph[currentNode]) {
            if (child !== parentNode) {
                bestValue = Math.max(bestValue, dfs(child, currentNode, depth + 1));
            }
        }
        return nodeValues[currentNode] = bestValue;
    } else {
        let bestValue = Number.MAX_SAFE_INTEGER;
        for (const child of graph[currentNode]) {
            if (child !== parentNode) {
                bestValue = Math.min(bestValue, dfs(child, currentNode, depth + 1));
            }
        }
        return nodeValues[currentNode] = bestValue;
    }
}

// Process the DFS
dfs(R, 0, 0);

// Handle queries
const Q = Number(input[N + L + 1]);
let result = "";

for (let i = 0; i < Q; i++) {
    const q = Number(input[N + L + 2 + i]);
    result += nodeValues[q] + "\n";
}

console.log(result.trim());

File: basic/minimaxtree/MinimaxTree.py
import sys

def dfs(current_node, parent_node, depth):
    if node_values[current_node] >= 0:
        return node_values[current_node]
    
    if depth % 2 == 0:
        best_value = 0
        for child in graph[current_node]:
            if child != parent_node:
                best_value = max(best_value, dfs(child, current_node, depth + 1))
        node_values[current_node] = best_value
        return best_value
    else:
        best_value = float('inf')
        for child in graph[current_node]:
            if child != parent_node:
                best_value = min(best_value, dfs(child, current_node, depth + 1))
        node_values[current_node] = best_value
        return best_value

# For faster input
input = sys.stdin.readline

# Read input
N, R = map(int, input().split())
node_values = [-1] * (N + 1)
graph = [[] for _ in range(N + 1)]

for _ in range(N - 1):
    u, v = map(int, input().split())
    graph[u].append(v)
    graph[v].append(u)

L = int(input())

for _ in range(L):
    leaf_id, leaf_value = map(int, input().split())
    node_values[leaf_id] = leaf_value

dfs(R, 0, 0)

Q = int(input())

for _ in range(Q):
    q = int(input())
    print(node_values[q])